File: dwl-ros-pkgs/dwl_msgs/src/dwl_msgs/WholeBodyStateInterface.py
#!/usr/bin/python
import roslib; roslib.load_manifest('dwl_msgs')

import dwl
from dwl_msgs.msg import WholeBodyState, BaseState, JointState, ContactState
import numpy as np
from xdg.Menu import tmp
from dwl.dwl import FloatingBaseSystem


class WholeBodyStateInterface():
    def __init__(self):
        return
    
    # This method fills the information of a WholeBodyState 
    # ROS message into a WholeBodyState object
    def reset(self, fbs):
        self.is_system = True
        self.fbs = fbs
    
    # Resets the floating-base system information
    def writeFromMessage(self, state, msg):
        # Setting the floating-base system DoF
        num_joints = len(msg.joints)
        num_contacts = len(msg.contacts)
        
        # Setting up the DoF
        state.setJointDoF(num_joints)

        # Getting the time
        state.setTime(msg.time)

        # Getting the base states
        base_pos = np.array([msg.base[dwl.LX].position,
                             msg.base[dwl.LY].position,
                             msg.base[dwl.LZ].position])
        base_rpy = np.array([msg.base[dwl.AX].position,
                             msg.base[dwl.AY].position,
                             msg.base[dwl.AZ].position])
        base_vel = np.array([msg.base[dwl.LX].velocity,
                             msg.base[dwl.LY].velocity,
                             msg.base[dwl.LZ].velocity])
        base_omega = np.array([msg.base[dwl.AX].velocity,
                               msg.base[dwl.AY].velocity,
                               msg.base[dwl.AZ].velocity])
        base_acc = np.array([msg.base[dwl.LX].velocity,
                             msg.base[dwl.LY].velocity,
                             msg.base[dwl.LZ].velocity])
        base_omegad = np.array([msg.base[dwl.AX].velocity,
                                msg.base[dwl.AY].velocity,
                                msg.base[dwl.AZ].velocity])
        state.setBasePosition_W(base_pos)
        state.setBaseRPY_W(base_rpy)
        state.setBaseVelocity_W(base_vel)
        state.setBaseAngularVelocity_W(base_omega)
        state.setBaseAcceleration_W(base_acc)
        state.setBaseAngularAcceleration_W(base_omegad)


        # Getting the joint states
        for i in range(num_joints):
            state.setJointPosition(msg.joints[i].position, i)
            state.setJointVelocity(msg.joints[i].velocity, i)
            state.setJointAcceleration(msg.joints[i].acceleration, i)
            state.setJointEffort(msg.joints[i].effort, i)

        # Getting the contact states
        for i in range(num_contacts):
            name = msg.contacts[i].name
            contact_pos = np.array([msg.contacts[i].position.x,
                                    msg.contacts[i].position.y,
                                    msg.contacts[i].position.z])
            contact_vel = np.array([msg.contacts[i].velocity.x,
                                    msg.contacts[i].velocity.y,
                                    msg.contacts[i].velocity.z])
            contact_acc = np.array([msg.contacts[i].acceleration.x,
                                    msg.contacts[i].acceleration.y,
                                    msg.contacts[i].acceleration.z])
            contact_wrc = np.array([msg.contacts[i].wrench.torque.x,
                                    msg.contacts[i].wrench.torque.y,
                                    msg.contacts[i].wrench.torque.z,
                                    msg.contacts[i].wrench.force.x,
                                    msg.contacts[i].wrench.force.y,
                                    msg.contacts[i].wrench.force.z])
            state.setContactPosition_B(name, contact_pos)
            state.setContactVelocity_B(name, contact_vel)
            state.setContactAcceleration_B(name, contact_acc)
            state.setContactWrench_B(name, contact_wrc)
    
    # This method fills the information of a WholeBodyState object 
    # into a WholeBodyState ROS message
    def writeToMessage(self, msg, state):
        if not is_system_:
            print("Warning: you cannot write the dwl_msg::WholeBodyState "
                "because it wasn't define the FloatingBaseSystem")
            return
        
        # Filling the time information
        msg.time = state.getTime()
        
        # Filling the base state
        msg.base
        for i in range(6):
            base_msg = BaseState()
            base_msg.id = i
            base_msg.position = state.base_pos[i]
            base_msg.velocity = state.base_vel[i]
            base_msg.acceleration = state.base_acc[i]
            msg.base.append(base_msg)
        
        fbs = FloatingBaseSystem()
        # Filling the joint state
        num_joints = state.getJointDoF()
        joint_names = self.fbs.getJointNames()
        for i in range(num_joints):
            joint_msg = JointState()
            joint_msg.name = joint_name[i]
            joint_msg.position = state.getJointPosition(i)
            joint_msg.velocity = state.getJointVelocity(i)
            joint_msg.acceleration = state.getJointAcceleration(i)
            joint_msg.effort = state.getJointEffort(i)
            msg.joint.append(joint_msg)
            
        # Filling the contact state
        for c in state.getContactPosition_B():
            contact_msg = ContactState()
            pos = state.getContactPosition_B(c)
            vel = state.getContactVelocity_B(c)
            acc = state.getContactAcceleration_B(c)
            wrc = state.getContactWrench_B(c)
            contact_msg.position.x = pos[dwl.X]
            contact_msg.position.y = pos[dwl.Y]
            contact_msg.position.z = pos[dwl.Z] 
            contact_msg.velocity.x = vel[dwl.X]
            contact_msg.velocity.y = vel[dwl.Y]
            contact_msg.velocity.z = vel[dwl.Z]
            contact_msg.acceleration.x = acc[dwl.X]
            contact_msg.acceleration.y = acc[dwl.Y]
            contact_msg.acceleration.z = acc[dwl.Z]
            contact_msg.wrench.torque.x = wrch[dwl.AX]
            contact_msg.wrench.torque.y = wrch[dwl.AY]
            contact_msg.wrench.torque.z = wrch[dwl.AZ]
            contact_msg.wrench.force.x = wrch[dwl.LX]
            contact_msg.wrench.force.y = wrch[dwl.LY]
            contact_msg.wrench.force.z = wrch[dwl.LZ]
            msg.contacts[c] = contact_msg

File: src/WholeBodyStateInterface.cpp
#include <dwl_msgs/WholeBodyStateInterface.h>


namespace dwl_msgs
{

WholeBodyStateInterface::WholeBodyStateInterface() : is_system_(false)
{

}


WholeBodyStateInterface::WholeBodyStateInterface(const dwl::model::FloatingBaseSystem& fbs) : is_system_(true)
{
	fbs_ = std::make_shared<dwl::model::FloatingBaseSystem>(fbs);
}


void WholeBodyStateInterface::reset(const dwl::model::FloatingBaseSystem& fbs)
{
	fbs_ = std::make_shared<dwl::model::FloatingBaseSystem>(fbs);
	is_system_ = true;
}


WholeBodyStateInterface::~WholeBodyStateInterface()
{

}


void WholeBodyStateInterface::writeToMessage(dwl_msgs::WholeBodyState& msg,
											 const dwl::WholeBodyState& state)
{
	if (!is_system_)
		printf(YELLOW "Warning: you cannot write the dwl_msg::WholeBodyState "
				"because it wasn't define the FloatingBaseSystem\n" COLOR_RESET);

	// Filling the time information
	msg.time = state.time;

	// Filling the base state
	if (fbs_->isFloatingBase()) {
		// Filling the data from the base state
		msg.base.name = fbs_->getFloatingBaseName();
		writetoMessage(msg.base.pose, state.getBaseSE3());
		writetoMessage(msg.base.velocity, state.getBaseVelocity_W());
		writetoMessage(msg.base.acceleration, state.getBaseAcceleration_W());
	}

	// Filling the joint state if there is information
	if (state.joint_pos.size() == fbs_->getJointDoF()) { //safety check
		msg.joints.resize(fbs_->getJointDoF());
		for (dwl::model::ElementId::const_iterator jnt_it = fbs_->getJoints().begin();
				jnt_it != fbs_->getJoints().end(); ++jnt_it) {
			std::string name = jnt_it->first;
			unsigned int id = jnt_it->second;

			msg.joints[id].name = name;
			msg.joints[id].position = state.joint_pos(id);
			msg.joints[id].velocity = state.joint_vel(id);
			msg.joints[id].acceleration = state.joint_acc(id);
			msg.joints[id].effort = state.joint_eff(id);
		}
	}

	// Filling the contact state if there is information
	// Note that planners can(not) generates contact information, at the same
	// time, controllers could(not) used it. If the controller need a missing
	// contact information, then it could computed from the joint trajectory
	unsigned int counter = 0;
	dwl::model::ElementId contact_links = fbs_->getEndEffectors();
	msg.contacts.resize(contact_links.size());
	for (dwl::model::ElementId::const_iterator it = contact_links.begin();
			it != contact_links.end(); ++it) {
		std::string name = it->first;

		// Filling the data from the contact state
		msg.contacts[counter].name = name;
		writetoMessage(msg.contacts[counter].pose, state.getContactSE3_B(name));
		writetoMessage(msg.contacts[counter].velocity, state.getContactVelocity_B(name));
		writetoMessage(msg.contacts[counter].acceleration, state.getContactAcceleration_B(name));
		writetoMessage(msg.contacts[counter].wrench, state.getContactWrench_B(name));

		// Incrementing the contact counter
		++counter;
	}
}


void WholeBodyStateInterface::writeToMessage(dwl_msgs::WholeBodyTrajectory& msg,
											 const dwl::WholeBodyTrajectory& traj)
{
	// Filling the trajectory
	unsigned int num_points = traj.size();
	msg.trajectory.resize(num_points);
	for (unsigned int i = 0; i < num_points; ++i)
		writeToMessage(msg.trajectory[i], traj[i]);
}


void WholeBodyStateInterface::writeFromMessage(dwl::WholeBodyState& state,
					  	  	  	  	  	  	   const dwl_msgs::WholeBodyState& msg)
{
	if (!is_system_)
		printf(YELLOW "Warning: you cannot write the dwl::WholeBodyState "
				"because it wasn't define the FloatingBaseSystem\n" COLOR_RESET);

	// Writing the time information
	state.time = msg.time;

	// Writing the base states
	writeFromMessage(state.base_pos, msg.base.pose);
	writeFromMessage(state.base_vel, msg.base.velocity);
	writeFromMessage(state.base_acc, msg.base.acceleration);

	// Sanity check: checking the size of the joint states
	if (state.joint_pos.size() != fbs_->getJointDoF())
		state.joint_pos = Eigen::VectorXd::Zero(fbs_->getJointDoF());
	if (state.joint_vel.size() != fbs_->getJointDoF())
		state.joint_vel = Eigen::VectorXd::Zero(fbs_->getJointDoF());
	if (state.joint_acc.size() != fbs_->getJointDoF())
		state.joint_acc = Eigen::VectorXd::Zero(fbs_->getJointDoF());
	if (state.joint_eff.size() != fbs_->getJointDoF())
		state.joint_eff = Eigen::VectorXd::Zero(fbs_->getJointDoF());

	// Writing the joint states
	unsigned num_joints = msg.joints.size();
	for (unsigned int i = 0; i < num_joints; ++i) {
		std::string name = msg.joints[i].name;
		unsigned int joint_id = fbs_->getJointId(name);

		state.joint_pos(joint_id) = msg.joints[i].position;
		state.joint_vel(joint_id) = msg.joints[i].velocity;
		state.joint_acc(joint_id) = msg.joints[i].acceleration;
		state.joint_eff(joint_id) = msg.joints[i].effort;
	}

	// Writing the contact states
//	Eigen::Vector3d contact_state;
	unsigned int num_contacts = msg.contacts.size();
	for (unsigned int i = 0; i < num_contacts; ++i) {
		// Getting the contact message
		dwl_msgs::ContactState contact_msg = msg.contacts[i];

		// Getting the contact name
		std::string name = contact_msg.name;

		dwl::SE3 pose;
		dwl::Motion vel, acc;
		dwl::Force wrch;

		// Getting the data from the message
		writeFromMessage(pose, contact_msg.pose);
		writeFromMessage(vel, contact_msg.velocity);
		writeFromMessage(acc, contact_msg.acceleration);
		writeFromMessage(wrch, contact_msg.wrench);

		// Filling the data inside the whole-body state
		state.setContactSE3_B(name, pose);
		state.setContactVelocity_B(name, vel);
		state.setContactAcceleration_B(name, acc);
		state.setContactWrench_B(name, wrch);
	}
}


void WholeBodyStateInterface::writeFromMessage(dwl::WholeBodyTrajectory& traj,
											   const dwl_msgs::WholeBodyTrajectory& msg)
{
	// Filling the trajectory
	unsigned int num_points = msg.trajectory.size();
	traj.resize(num_points);
	for (unsigned int i = 0; i < num_points; ++i)
		writeFromMessage(traj[i], msg.trajectory[i]);
}


void WholeBodyStateInterface::writetoMessage(geometry_msgs::Pose& msg,
											 const dwl::SE3& state)
{
	const Eigen::Vector7d& pose = state.toVector();

	// Filling the pose
	msg.position.x = pose(dwl::rbd::LX_Q);
	msg.position.y = pose(dwl::rbd::LY_Q);
	msg.position.z = pose(dwl::rbd::LZ_Q);
	msg.orientation.x = pose(dwl::rbd::AX_Q);
	msg.orientation.y = pose(dwl::rbd::AY_Q);
	msg.orientation.z = pose(dwl::rbd::AZ_Q);
	msg.orientation.w = pose(dwl::rbd::AW_Q);
}


void WholeBodyStateInterface::writetoMessage(geometry_msgs::Twist& msg,
											 const dwl::Motion& state)
{
	const Eigen::Vector6d& twist = state.toVector();

	// Filling the pose
	msg.linear.x = twist(dwl::rbd::LX_V);
	msg.linear.y = twist(dwl::rbd::LY_V);
	msg.linear.z = twist(dwl::rbd::LZ_V);
	msg.angular.x = twist(dwl::rbd::AX_V);
	msg.angular.y = twist(dwl::rbd::AY_V);
	msg.angular.z = twist(dwl::rbd::AZ_V);
}


void WholeBodyStateInterface::writetoMessage(geometry_msgs::Wrench& msg,
											 const dwl::Force& state)
{
	const Eigen::Vector6d& wrench = state.toVector();

	// Filling the pose
	msg.force.x = wrench(dwl::rbd::LX_V);
	msg.force.y = wrench(dwl::rbd::LY_V);
	msg.force.z = wrench(dwl::rbd::LZ_V);
	msg.torque.x = wrench(dwl::rbd::AX_V);
	msg.torque.y = wrench(dwl::rbd::AY_V);
	msg.torque.z = wrench(dwl::rbd::AZ_V);
}


void WholeBodyStateInterface::writeFromMessage(dwl::SE3& state,
											   const geometry_msgs::Pose& msg)
{
	// Filling the message in the state
	state.setTranslation(Eigen::Vector3d(msg.position.x,
										 msg.position.y,
										 msg.position.z));
	state.setQuaternion(Eigen::Vector4d(msg.orientation.x,
										msg.orientation.y,
										msg.orientation.z,
										msg.orientation.w));
}


void WholeBodyStateInterface::writeFromMessage(dwl::Motion& state,
					  	  	  	  	  	  	   const geometry_msgs::Twist& msg)
{
	// Filling the message in the state
	state.setLinear(Eigen::Vector3d(msg.linear.x,
									msg.linear.y,
									msg.linear.z));
	state.setAngular(Eigen::Vector3d(msg.angular.x,
									 msg.angular.y,
									 msg.angular.z));
}


void WholeBodyStateInterface::writeFromMessage(dwl::Force& state,
					  	  	  	  	  	  	   geometry_msgs::Wrench& msg)
{
	// Filling the message in the state
	state.setLinear(Eigen::Vector3d(msg.force.x,
									msg.force.y,
									msg.force.z));
	state.setAngular(Eigen::Vector3d(msg.torque.x,
									 msg.torque.y,
									 msg.torque.z));
}

} //@namespace dwl_msgs
